### PR TITLE
Adds an option to locally fetch site website content

### PR DESCRIPTION
**Pull request by agallou**
Some websites like https://www.arretsurimages.net are Single Page Applications behind a Paywall.
Those websites cannot be fetched by wallabag using the credentials API.

This PR adds an option where you can list the websites that you want to fetch locally
Capture d’écran du 2021-12-05 17-17-57

The tab's page DOM will be used as content and the tab's title will be used as title.

### CODE REVIEW

**Comment by Simounet** [comment]
Hi Adrien,
First of all, thanks for this useful feature. Could you rebase please? You should have some CS linting to do. Let me know if I can do anything to help you.

**Comment by agallou** [comment]
@Simounet i've just rebased and fixed most of the linter errors. 
There is still two errors : 
```
/sources/wallabagger/js/options.js
  145:45  error  document.body() is not supported in Firefox 52  compat/compat
  147:9   error  document.body() is not supported in Firefox 52  compat/compat
```
Do you know what's the best way to handle them ?

File: wallabagger/js/options.js [hunk 1]
```diff
@@ -55,6 +56,7 @@ OptionsController.prototype = {
     loadFromFileButton: null,
     openFileDialog: null,
     clearButton: null,
+    SitesToFetchLocally: null,
```
Comment (Simounet) [inline wallabagger/js/options.js:59]: I'm sorry but could you use camelCame here please?

**Comment by agallou** [comment]
@Simounet i've just rebased and fixed most of the linter errors. 
There is still two errors : 
```
/sources/wallabagger/js/options.js
  145:45  error  document.body() is not supported in Firefox 52  compat/compat
  147:9   error  document.body() is not supported in Firefox 52  compat/compat
```
Do you know what's the best way to handle them ?

**Review by Simounet — CHANGES_REQUESTED** [review]
Thanks a lot!

**Comment by Simounet** [comment]
> @Simounet i've just rebased and fixed most of the linter errors. There is still two errors :
> 
> ```
> /sources/wallabagger/js/options.js
>   145:45  error  document.body() is not supported in Firefox 52  compat/compat
>   147:9   error  document.body() is not supported in Firefox 52  compat/compat
> ```
> 
> Do you know what's the best way to handle them ?

That's weird because I fixed it on the master branch with this commit: https://github.com/wallabag/wallabagger/commit/32cb230d3c3e3fdb875ed33e70e8a2686528cd5c

File: wallabagger/js/popup.js [hunk 2]
```diff
@@ -609,7 +609,24 @@ PopupController.prototype = {
             this.cardTitle.textContent = tab.title;
             this.entryUrl.textContent = /(\w+:\/\/)([^/]+)\/(.*)/.exec(tab.url)[2];
             this.enableTagsInput();
-            this.port.postMessage({ request: 'save', tabUrl: tab.url });
+
+            const port = this.port;
```
[the remark this answers is not in the document]
Comment (agallou) [reply wallabagger/js/popup.js:613]: without that, if I use directly this.port, I get this error : `Uncaught (in promise) Error: this.port is undefined`.

**Comment by agallou** [comment]
> That's weird because I fixed it on the master branch with this commit: https://github.com/wallabag/wallabagger/commit/32cb230d3c3e3fdb875ed33e70e8a2686528cd5c

Sorry, I did not fetched the latest changes before rebasing. That's fixed.

I've fixed the typos/conventions errors, improved the filtering of the pages and used an object to avoid duplication of the function.

If all of that is ok for you I can then squash those changes.

**Comment by nicosomb** [comment]
> The tab's page DOM will be used as content

Do you send the tab's page DOM to wallabag server? (to clean it)

**Review by Simounet — CHANGES_REQUESTED** [review]
I think we should be done after that. Thanks again for your time.

**Comment by agallou** [comment]
@nicosomb yes, the innerHTML of the body is retrived here : https://github.com/wallabag/wallabagger/pull/300/files#diff-bfbd252c728bc45cc06ffc1d7f8eb7b9d5489941ad1c9183d8708245cc257e1eR624
then it is send to wallabag's API via the content parameter here : https://github.com/wallabag/wallabagger/pull/300/files#diff-45a26550cb91696cb8ce6bab9815b6ed166a4b3901d68c64354d79887a171501R201

**Comment by agallou** [comment]
@Simounet i've just applied changes about the function/this and variable naming, then rebased/squashed the branch.

**Comment by Simounet** [comment]
@agallou Everything looks good to me. Thanks a lot!
@nicosomb We are posting the content to `/api/entries.json`. Is it enough for wallabag to clean it?

**Review by Simounet — APPROVED** [review]

**Comment by nicosomb** [comment]
I think it's enough. 

If I understand, you only do this new thing for selected websites?

**Comment by agallou** [comment]
> I think it's enough.
> 
> If I understand, you only do this new thing for selected websites?

Yes, its only done for the sites listed in the configuration. It's usefull if the site is and SPA and/or if the site is behind a paywall that is not managed/can't be managed with wallabag's site credentials.

**Comment by Simounet** [comment]
I added a commit to bypass the cache for your local fetch and updated the translations. It's now on the `master` branch. Thanks a lot for this feature!

**Comment by nicosomb** [comment]
But why this feature is not enabled for all the websites?

**Comment by Simounet** [comment]
It could be but at the moment, submitting an entry to wallabag through the right click context menu on a link action is not supported. You have to open it on a new tab before.

**Pull request closed without merging**